# DevRev Search Dataset

Loading and exploring the `devrev/search` dataset from Hugging Face.

**Dataset Structure:**
- `annotated_queries` — Queries paired with annotated (golden) article chunks
- `knowledge_base` — Article chunks from DevRev's customer-facing support documentation
- `test_queries` — Held-out queries used for evaluation

In [4]:
from datasets import load_dataset
import pandas as pd
from dotenv import load_dotenv
load_dotenv()

True

## 1. Load Annotated Queries
Queries paired with annotated (golden) article chunks for training/validation.

In [5]:
# Load annotated queries
annotated_queries = load_dataset("devrev/search", "annotated_queries")
print(annotated_queries)

DatasetDict({
    train: Dataset({
        features: ['query_id', 'query', 'retrievals'],
        num_rows: 291
    })
})


In [6]:
# Convert to DataFrame and display
annotated_df = annotated_queries["train"].to_pandas()
annotated_df.head()

,query_id,query,retrievals
0,0ae94217-c6a0-4895-83a2-841a95f01637,create DevRev ticket from Microsoft Teams,"[{'id': 'ART-4216_KNOWLEDGE_NODE-26', 'text': ..."
1,d0b209b3-6cea-46d8-bfac-bd0e286ea21b,workflow builder auto close ticket after 48 ho...,"[{'id': 'ART-2012_KNOWLEDGE_NODE-24', 'text': ..."
2,40c1aa6f-cd21-46ab-8f6f-76fdc267b584,automated reminder to customer ticket will be ...,"[{'id': 'ART-3068_KNOWLEDGE_NODE-24', 'text': ..."
3,e47d883f-b712-4f98-bd06-14ade143e3c2,connect Bitbucket account to DevRev account,"[{'id': 'ART-2030_KNOWLEDGE_NODE-27', 'text': ..."
4,2e6f9413-15ac-4974-a380-7aa22fc98a61,use of workflows in DevRev,"[{'id': 'ART-1961_KNOWLEDGE_NODE-28', 'text': ..."


In [7]:
# Sample a single annotated query example
annotated_queries["train"][0]

{'query_id': '0ae94217-c6a0-4895-83a2-841a95f01637',
 'query': 'create DevRev ticket from Microsoft Teams',
 'retrievals': [{'id': 'ART-4216_KNOWLEDGE_NODE-26',
   'text': 'DevRev Object | Sync to DevRev |\\n| --- | --- | --- |\\n| Plan | Parts | \\xe2\\x9c\\x85 |\\n| User | Identity/DevUser | \\xe2\\x9c\\x85 |\\n| Channel | Chat | \\xe2\\x9c\\x85 |\\n| Attachments in Message/Thread/Task | Artifacts on Article | \\xe2\\x9c\\x85 |\\n| Message | Comment | \\xe2\\x9c\\x85 |\\n| Thread | Comment | \\xe2\\x9c\\x85 |\\n| Task | Issue/Ticket | \\xe2\\x9c\\x85 |\\n\\nImporting from Microsoft Teams\\n------------------------------\\n\\nFollow the steps below to import from Microsoft Teams:\\n\\n1. Go to',
   'title': 'Microsoft Teams AirSync | AirSync | Snap-ins | DevRev'},
  {'id': 'ART-4216_KNOWLEDGE_NODE-29',
   'text': 'with many\\nattachments. DevRev honors the Microsoft Graph API rate limits and back-off and resumes automatically.\\n\\nPost import options\\n-------------------\\n\\nAfter 

## 2. Load Test Queries
Held-out queries used for evaluation.

In [8]:
# Load test queries
test_queries = load_dataset("devrev/search", "test_queries")
print(test_queries)

DatasetDict({
    test: Dataset({
        features: ['query_id', 'query'],
        num_rows: 92
    })
})


In [9]:
# Convert to DataFrame and display
test_df = test_queries["test"].to_pandas()
test_df.head()

,query_id,query
0,a97f93d2-410a-431f-ae9a-1e23ed35d74c,end customer organization name not appearing i...
1,7dd7e2b4-9349-4535-8007-1d706e0fabff,Android SDK session generated with Unknown user
2,4bc92187-cdaa-4c20-b189-abd1672e5a71,email reply received on wrong ticket
3,4d9878e8-f746-4df5-8bf6-f9444989b385,manage access and privileges in DevRev
4,483151ec-aff4-4569-b3df-651f578b61d8,SSO setup SAML IDP metadata connection string ...


In [10]:
# Sample a single test query example
test_queries["test"][0]

{'query_id': 'a97f93d2-410a-431f-ae9a-1e23ed35d74c',
 'query': 'end customer organization name not appearing in ticket or conversation'}

## 3. Load Knowledge Base
Article chunks from DevRev's customer-facing support documentation.

In [11]:
# Load knowledge base
knowledge_base = load_dataset("devrev/search", "knowledge_base")
print(knowledge_base)

DatasetDict({
    corpus: Dataset({
        features: ['id', 'text', 'title'],
        num_rows: 65224
    })
})


In [12]:
# Convert to DataFrame and display
knowledge_df = knowledge_base["corpus"].to_pandas()
knowledge_df.head()

,id,text,title
0,ART-17711_KNOWLEDGE_NODE-0,b'We ran into a case where an AirSync was star...,Sync fails when original sync owners loses per...
1,ART-17711_KNOWLEDGE_NODE-1,access.\n\nOnce Person A was re-added with the...,Sync fails when original sync owners loses per...
2,ART-17650_KNOWLEDGE_NODE-0,"b""American cybersecurity leader unifies securi...",American cybersecurity leader unifies security...
3,ART-17650_KNOWLEDGE_NODE-1,DevRev\n======================================...,American cybersecurity leader unifies security...
4,ART-17650_KNOWLEDGE_NODE-2,solutions help organisations build and deploy ...,American cybersecurity leader unifies security...


In [13]:
# Sample a single knowledge base chunk
knowledge_base["corpus"][0]

{'id': 'ART-17711_KNOWLEDGE_NODE-0',
 'text': "b'We ran into a case where an AirSync was started by one person (Person A) and later failed. Another user (Person B) tried to click Retry, but it didn\\xe2\\x80\\x99t work. The logs showed 401 and 403 errors in communication between the snap-in and the snap-in manager.\\n\\nIt turned out that AirSync assigns the sync owner to whoever started it. Since Person A had been removed from the org or lost permissions, the retry failed \\xe2\\x80\\x94 the system still expected the original owner to have valid",
 'title': 'Sync fails when original sync owners loses permissions'}

## 4. Dataset Summary

In [14]:
print("=" * 60)
print("DevRev Search Dataset Summary")
print("=" * 60)
print(f"\nAnnotated Queries:")
print(annotated_queries)
print(f"\nTest Queries:")
print(test_queries)
print(f"\nKnowledge Base:")
print(knowledge_base)
print("\n" + "=" * 60)

DevRev Search Dataset Summary

Annotated Queries:
DatasetDict({
    train: Dataset({
        features: ['query_id', 'query', 'retrievals'],
        num_rows: 291
    })
})

Test Queries:
DatasetDict({
    test: Dataset({
        features: ['query_id', 'query'],
        num_rows: 92
    })
})

Knowledge Base:
DatasetDict({
    corpus: Dataset({
        features: ['id', 'text', 'title'],
        num_rows: 65224
    })
})



---
## 5. Index Knowledge Base with FAISS

Using OpenAI text-embedding-3-small and FAISS for similarity search.

In [15]:
from openai import OpenAI
import faiss
import numpy as np
from tqdm import tqdm
import time
import os
import json
import pickle

In [16]:
# Initialize OpenAI client
# Set your API key as an environment variable: export OPENAI_API_KEY="your-key-here"
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
if not OPENAI_API_KEY:
    raise ValueError("Please set OPENAI_API_KEY environment variable")

client = OpenAI(api_key=OPENAI_API_KEY)

MODEL_ID = "text-embedding-3-small"  # 1536 dimensions
print(f"Using model: {MODEL_ID}")
print(f"Provider: OpenAI")

Using model: text-embedding-3-small
Provider: OpenAI


In [17]:
def get_embedding(text: str) -> np.ndarray:
    """Get embedding for a single text using OpenAI text-embedding-3."""
    response = client.embeddings.create(
        model=MODEL_ID,
        input=text
    )
    return np.array(response.data[0].embedding)

In [18]:
def get_embeddings_batch(texts: list) -> np.ndarray:
    """Get embeddings for a batch of texts using OpenAI text-embedding-3."""
    response = client.embeddings.create(
        model=MODEL_ID,
        input=texts
    )
    return np.array([d.embedding for d in response.data])

In [19]:
from sentence_transformers import SentenceTransformer
import numpy as np

model = SentenceTransformer("all-MiniLM-L6-v2")  # free, runs locally

def get_embedding(text: str) -> np.ndarray:
    return model.encode(text)

def get_embeddings_batch(texts: list) -> np.ndarray:
    return model.encode(texts)

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 1907.55it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [20]:
# Test the embedding API
test_embedding = get_embedding("This is a test sentence.")
print(f"Test embedding shape: {test_embedding.shape}")
print(f"Embedding dimension: {len(test_embedding)}")
print(f"First 5 values: {test_embedding[:5]}")

Test embedding shape: (384,)
Embedding dimension: 384
First 5 values: [0.08429643 0.05795369 0.00449337 0.1058211  0.00708341]


In [21]:
# Prepare documents: concatenate title with text
corpus = knowledge_base["corpus"]

documents = []
doc_ids = []
doc_titles = []
doc_texts = []

for item in tqdm(corpus, desc="Preparing documents"):
    # Concatenate title and text
    doc_text = f"{item['title']}\n\n{item['text']}"
    documents.append(doc_text)
    doc_ids.append(item['id'])
    doc_titles.append(item['title'])
    doc_texts.append(item['text'])

print(f"\nTotal documents: {len(documents):,}")
print(f"\nSample document:")
print(documents[0][:500])

Preparing documents: 100%|██████████| 65224/65224 [00:00<00:00, 118407.86it/s]


Total documents: 65,224

Sample document:
Sync fails when original sync owners loses permissions

b'We ran into a case where an AirSync was started by one person (Person A) and later failed. Another user (Person B) tried to click Retry, but it didn\xe2\x80\x99t work. The logs showed 401 and 403 errors in communication between the snap-in and the snap-in manager.\n\nIt turned out that AirSync assigns the sync owner to whoever started it. Since Person A had been removed from the org or lost permissions, the retry failed \xe2\x80\x94 the s


In [22]:
def get_all_embeddings(texts, batch_size=100, max_retries=3):
    """Get embeddings for all texts using OpenAI batch API."""
    all_embeddings = []
    
    for i in tqdm(range(0, len(texts), batch_size), desc="Generating embeddings"):
        batch = texts[i:i + batch_size]
        
        # Truncate texts if too long (OpenAI has ~8K token limit)
        batch = [text[:8000] if len(text) > 8000 else text for text in batch]
        
        retries = 0
        while retries < max_retries:
            try:
                embeddings = get_embeddings_batch(batch)
                all_embeddings.append(embeddings)
                break
            except Exception as e:
                retries += 1
                if retries >= max_retries:
                    print(f"Error embedding batch {i}: {e}")
                    all_embeddings.append(np.zeros((len(batch), 1536)))
                else:
                    print(f"Retry {retries} for batch {i}...")
                    time.sleep(2)
        
        time.sleep(0.1)  # Rate limiting
    
    return np.vstack(all_embeddings)

In [23]:
# Generate embeddings for all documents
print("Generating embeddings for knowledge base...")
print(f"Total documents: {len(documents):,}")
print("Using batch processing for efficiency...")

# For testing, use a subset (uncomment for all documents)
# documents_to_embed = documents[:100]  # Test with first 100 docs
documents_to_embed = documents  # All documents

embeddings = get_all_embeddings(documents_to_embed, batch_size=1000)
print(f"\nEmbeddings shape: {embeddings.shape}")

# Save embeddings
np.save("embeddings.npy", embeddings)
print("Embeddings saved to embeddings.npy")

Generating embeddings for knowledge base...
Total documents: 65,224
Using batch processing for efficiency...


Generating embeddings: 100%|██████████| 66/66 [01:41<00:00,  1.54s/it]


Embeddings shape: (65224, 384)
Embeddings saved to embeddings.npy


In [24]:
# Load embeddings if already saved
# embeddings = np.load("embeddings.npy")
# print(f"Loaded embeddings shape: {embeddings.shape}")

In [25]:
# Create FAISS index
embedding_dim = embeddings.shape[1]
print(f"Creating FAISS index with dimension: {embedding_dim}")

# Normalize embeddings for cosine similarity
embeddings_normalized = embeddings.copy().astype('float32')
faiss.normalize_L2(embeddings_normalized)

# Create the index using IndexFlatIP for inner product (cosine similarity with normalized vectors)
index = faiss.IndexFlatIP(embedding_dim)
index.add(embeddings_normalized)

print(f"Index created with {index.ntotal:,} vectors")

: 

In [25]:
# Save the index and document mapping
INDEX_DIR = "faiss_index"
os.makedirs(INDEX_DIR, exist_ok=True)

# Save FAISS index
faiss.write_index(index, os.path.join(INDEX_DIR, "knowledge_base.index"))

# Save document mapping
with open(os.path.join(INDEX_DIR, "doc_mapping.pkl"), "wb") as f:
    pickle.dump({
        "doc_ids": doc_ids,
        "documents": documents,
        "doc_titles": doc_titles,
        "doc_texts": doc_texts
    }, f)

print(f"✓ Index saved to {INDEX_DIR}/knowledge_base.index")
print(f"✓ Document mapping saved to {INDEX_DIR}/doc_mapping.pkl")

✓ Index saved to faiss_index/knowledge_base.index
✓ Document mapping saved to faiss_index/doc_mapping.pkl


## 6. Search the Knowledge Base

In [26]:
def search(query: str, k: int = 5):
    """Search the knowledge base for relevant documents."""
    query_embedding = get_embedding(query).astype('float32')
    query_embedding = query_embedding.reshape(1, -1)
    faiss.normalize_L2(query_embedding)
    
    scores, indices = index.search(query_embedding, k)
    
    results = []
    for i, (score, idx) in enumerate(zip(scores[0], indices[0])):
        results.append({
            "rank": i + 1,
            "score": float(score),
            "id": doc_ids[idx],
            "title": doc_titles[idx],
            "text": doc_texts[idx]
        })
    
    return results

In [27]:
# Test search with a sample query
query = "How do I set up AirSync?"
results = search(query, k=5)

print(f"Query: {query}")
print("=" * 60)

for r in results:
    print(f"\n[Rank {r['rank']}] Score: {r['score']:.4f}")
    print(f"Doc ID: {r['id']}")
    print(f"Title: {r['title']}")
    print(f"Text: {r['text'][:300]}...")
    print("-" * 40)

Query: How do I set up AirSync?

[Rank 1] Score: 0.6888
Doc ID: ART-17212_KNOWLEDGE_NODE-5
Title: AirSync snap-in template | DevRev | Docs
Text: snap-in. Inside, the `functions/` subfolder includes the logic for retrieving external sync units, extracting data from and loading data to external systems, and mapping between them.\n\nWas this page helpful?\n\nYesNo\n\n[Previous](/airsync/getting-started)[#### Local development\n\nNext](/airsync/...
----------------------------------------

[Rank 2] Score: 0.6721
Doc ID: ART-2045_KNOWLEDGE_NODE-29
Title: AirSync | Snap-ins | DevRev
Text: Setting up a new AirSync\n\n![]()\n\nFor best results, AirSyncs should be done using an administrator account on the external source. This ensures all necessary permissions are available to complete the import successfully.\n\nWhether you want to perform only a one-time import or set up an ongoing s...
----------------------------------------

[Rank 3] Score: 0.6712
Doc ID: ART-2047_KNOWLEDGE_NODE-29
Title:

---
## 7. Run Evaluation on Test Queries

Run search against all test queries and save results in the same format as annotated_queries.

In [28]:
# Run search on all test queries - DIRECT VERSION (no dependency on search function)
TOP_K = 10  # Number of retrievals per query

# Get corpus data for lookups
corpus_data = knowledge_base["corpus"]

test_results = []

for item in tqdm(test_queries["test"], desc="Processing test queries"):
    query_id = item["query_id"]
    query = item["query"]
    
    # Get query embedding directly
    query_embedding = get_embedding(query).astype('float32').reshape(1, -1)
    faiss.normalize_L2(query_embedding)
    
    # Search FAISS index
    scores, indices = index.search(query_embedding, TOP_K)
    
    # Format retrievals using corpus data directly
    retrievals = []
    for idx in indices[0]:
        doc = corpus_data[int(idx)]
        retrievals.append({
            "id": doc["id"],
            "text": doc["text"],
            "title": doc["title"]
        })
    
    test_results.append({
        "query_id": query_id,
        "query": query,
        "retrievals": retrievals
    })

print(f"\nProcessed {len(test_results)} test queries")

Processing test queries: 100%|██████████| 92/92 [00:00<00:00, 106.21it/s]


Processed 92 test queries


In [29]:
# Preview a sample result
import json
print("Sample result:")
print(json.dumps(test_results[0], indent=2, default=str)[:1500])

Sample result:
{
  "query_id": "a97f93d2-410a-431f-ae9a-1e23ed35d74c",
  "query": "end customer organization name not appearing in ticket or conversation",
  "retrievals": [
    {
      "id": "ART-1978_KNOWLEDGE_NODE-44",
      "text": "URL and not on the support portal of some other customer.\\n* Customer admin isn't able to see all the tickets of the organization.\\n\\n  + This could happen if the customer isn't logged in on the correct URL. If the customer is logged in on the correct URL, then check if there are any tickets that are reported by the other customers in that organization or not. Check if the customer is added as a customer admin or not by logging in to your DevRev application.\\n* You are not able to add customer",
      "title": "Customer portal | Computer for Support Teams | DevRev"
    },
    {
      "id": "ART-6174_KNOWLEDGE_NODE-27",
      "text": "happens automatically:\\n\\n* The original conversation moves to *Archived* stage and cannot be reopened.\\n* A new t

In [30]:
# Save results to JSON file
OUTPUT_FILE = "test_queries_results.json"

with open(OUTPUT_FILE, "w") as f:
    json.dump(test_results, f, indent=2)

print(f"✓ Results saved to {OUTPUT_FILE}")
print(f"  - {len(test_results)} queries")
print(f"  - {TOP_K} retrievals per query")

✓ Results saved to test_queries_results.json
  - 92 queries
  - 10 retrievals per query


In [31]:
# Also save as a parquet file for easier loading
results_df = pd.DataFrame(test_results)
results_df.to_parquet("test_queries_results.parquet", index=False)
print("✓ Results also saved to test_queries_results.parquet")

✓ Results also saved to test_queries_results.parquet


In [32]:
# Display results summary
print("=" * 60)
print("Test Queries Results Summary")
print("=" * 60)
print(f"Total queries: {len(test_results)}")
print(f"Retrievals per query: {TOP_K}")
print(f"\nOutput files:")
print(f"  - test_queries_results.json")
print(f"  - test_queries_results.parquet")
print("\nFormat matches annotated_queries structure:")
print("  - query_id: string")
print("  - query: string")
print("  - retrievals: list of {id, text, title}")

Test Queries Results Summary
Total queries: 92
Retrievals per query: 10

Output files:
  - test_queries_results.json
  - test_queries_results.parquet

Format matches annotated_queries structure:
  - query_id: string
  - query: string
  - retrievals: list of {id, text, title}


## 8. Load Saved Index (Optional)
Use this to load a previously saved index without re-embedding.

In [33]:
# Load saved index and mapping
# INDEX_DIR = "faiss_index"
# index = faiss.read_index(os.path.join(INDEX_DIR, "knowledge_base.index"))
# with open(os.path.join(INDEX_DIR, "doc_mapping.pkl"), "rb") as f:
#     mapping = pickle.load(f)
#     doc_ids = mapping["doc_ids"]
#     documents = mapping["documents"]
#     doc_titles = mapping["doc_titles"]
#     doc_texts = mapping["doc_texts"]
# print(f"Loaded index with {index.ntotal:,} vectors")

---
## 9. Enhanced Search Pipeline

**Target: match Rank 3 open-source entry — `snowflake-arctic-embed-l-v2.0-bm25-zerank-1-small`**

| Stage | What | Why |
|-------|------|-----|
| **Text Cleaning** | Remove `b'...'` byte artifacts, normalize unicode | Cleaner text → better embeddings & BM25 |
| **Stronger Embeddings** | `snowflake-arctic-embed-l-v2.0` (1024-dim, multilingual) | Exact model used by Rank 3 open-source entry |
| **Hybrid Retrieval** | Dense + BM25 with Reciprocal Rank Fusion | Catches both semantic and keyword matches |
| **Cross-Encoder Reranking** | `zeroentropy/zerank-1-small` (Apache 2.0) | Exact reranker used by all Top-4 entries |

**Leaderboard gap (all values in %):**

| Metric | You (current) | Rank 3 open-source | Gap |
|--------|:---:|:---:|:---:|
| Precision@10 | 16.56 | 26.63 | -10.07 pts |
| Recall@10 | 20.98 | 36.09 | -15.11 pts |

> **Note on metrics:** The leaderboard evaluates on `test_queries` (92 held-out queries submitted to DevRev). Your local evaluation uses `annotated_queries` as a proxy. Local numbers will differ slightly from leaderboard numbers but track the same trend.

In [1]:
import subprocess
subprocess.check_call(["pip", "install", "-q", "rank_bm25"])

import re, ast, torch
from rank_bm25 import BM25Okapi
from sentence_transformers import SentenceTransformer, CrossEncoder
from collections import defaultdict

# Auto-select best available device: MPS (Apple Silicon) > CUDA > CPU
if torch.backends.mps.is_available():
    DEVICE = "mps"
elif torch.cuda.is_available():
    DEVICE = "cuda"
else:
    DEVICE = "cpu"

print(f"Device selected: {DEVICE.upper()}")
if DEVICE == "mps":
    print("Apple Silicon MPS enabled — GPU acceleration active.")

/Users/macbookpro/Desktop/DevRev_search/.venv/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Device selected: MPS
Apple Silicon MPS enabled — GPU acceleration active.


### 9.1 Evaluation Metrics
Measure Recall@K, Precision@K, and MRR against annotated golden retrievals.

In [2]:
def evaluate_retrieval(predictions, ground_truth, k_values=[1, 3, 5, 10]):
    """
    Evaluate retrieval quality: Recall@K, Precision@K, MRR.
    Returns metrics in PERCENTAGE form (0-100) to match leaderboard scale.
    """
    gt_map = {g["query_id"]: set(r["id"] for r in g["retrievals"]) for g in ground_truth}

    metrics = {}
    for k in k_values:
        recalls, precisions = [], []
        for pred in predictions:
            golden_ids = gt_map.get(pred["query_id"], set())
            if not golden_ids:
                continue
            predicted_ids = [r["id"] for r in pred["retrievals"][:k]]
            hits = sum(1 for pid in predicted_ids if pid in golden_ids)
            recalls.append(hits / len(golden_ids))
            precisions.append(hits / k)
        metrics[f"Recall@{k}"]    = 100 * sum(recalls)    / len(recalls)    if recalls    else 0
        metrics[f"Precision@{k}"] = 100 * sum(precisions) / len(precisions) if precisions else 0

    mrrs = []
    for pred in predictions:
        golden_ids = gt_map.get(pred["query_id"], set())
        if not golden_ids:
            continue
        for rank, r in enumerate(pred["retrievals"], 1):
            if r["id"] in golden_ids:
                mrrs.append(1.0 / rank)
                break
        else:
            mrrs.append(0.0)
    metrics["MRR"] = 100 * sum(mrrs) / len(mrrs) if mrrs else 0

    return metrics

print("Evaluation function ready (metrics in %, matching leaderboard scale).")

Evaluation function ready (metrics in %, matching leaderboard scale).


In [3]:
# Evaluate current baseline (all-MiniLM-L6-v2 + FAISS)
print("Evaluating baseline (all-MiniLM-L6-v2 + FAISS flat)...")

corpus_data = knowledge_base["corpus"]
ground_truth = [dict(item) for item in annotated_queries["train"]]

baseline_results = []
for item in tqdm(annotated_queries["train"], desc="Baseline eval"):
    query_emb = get_embedding(item["query"]).astype('float32').reshape(1, -1)
    faiss.normalize_L2(query_emb)
    scores, indices = index.search(query_emb, 10)

    retrievals = []
    for idx in indices[0]:
        doc = corpus_data[int(idx)]
        retrievals.append({"id": doc["id"], "text": doc["text"], "title": doc["title"]})
    baseline_results.append({
        "query_id": item["query_id"],
        "query": item["query"],
        "retrievals": retrievals,
    })

baseline_metrics = evaluate_retrieval(baseline_results, ground_truth)
print("\n=== BASELINE Performance (%, matching leaderboard scale) ===")
for k, v in sorted(baseline_metrics.items()):
    print(f"  {k:15s}: {v:6.2f}%")

Evaluating baseline (all-MiniLM-L6-v2 + FAISS flat)...


NameError: name 'knowledge_base' is not defined

### 9.2 Text Cleaning
Many knowledge base entries have raw byte-string wrappers (`b'...'`) and escaped unicode. Clean these for better embedding and BM25 quality.

In [37]:
def clean_text(text: str) -> str:
    """Remove byte-string wrappers, decode escaped unicode, normalize whitespace."""
    if isinstance(text, str) and (text.startswith("b'") or text.startswith('b"')):
        try:
            evaluated = ast.literal_eval(text)
            if isinstance(evaluated, bytes):
                text = evaluated.decode("utf-8", errors="replace")
        except (ValueError, SyntaxError):
            text = text[2:-1]

    text = text.replace("\\n", "\n")
    text = re.sub(r"\\x[0-9a-fA-F]{2}", " ", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r" {2,}", " ", text)
    return text.strip()


cleaned_documents = []
cleaned_texts = []
for title, raw_text in zip(doc_titles, doc_texts):
    ct = clean_text(raw_text)
    cleaned_texts.append(ct)
    cleaned_documents.append(f"{title}\n\n{ct}")

print(f"Cleaned {len(cleaned_documents):,} documents")
print(f"\nBEFORE: {documents[0][:200]}")
print(f"\nAFTER:  {cleaned_documents[0][:200]}")

Cleaned 65,224 documents

BEFORE: Sync fails when original sync owners loses permissions

b'We ran into a case where an AirSync was started by one person (Person A) and later failed. Another user (Person B) tried to click Retry, but i

AFTER:  Sync fails when original sync owners loses permissions

We ran into a case where an AirSync was started by one person (Person A) and later failed. Another user (Person B) tried to click Retry, but it 


### 9.3 Stronger Dense Embedding Model
Replace `all-MiniLM-L6-v2` (384-dim) with `Snowflake/snowflake-arctic-embed-l-v2.0` (1024-dim) — the exact embedding model used by the Rank 3 open-source leaderboard entry.

Snowflake Arctic Embed v2 uses `prompt_name="query"` for query encoding (vs document encoding), which is handled automatically by the SentenceTransformers integration.

In [39]:
from sentence_transformers import SentenceTransformer

DENSE_MODEL_NAME = "Snowflake/snowflake-arctic-embed-l-v2.0"
print(f"Loading {DENSE_MODEL_NAME} on {DEVICE.upper()}...")
dense_model = SentenceTransformer(DENSE_MODEL_NAME, device=DEVICE)
print(f"Embedding dimension: {dense_model.get_sentence_embedding_dimension()}")
print(f"Model device: {next(dense_model.parameters()).device}")


def get_dense_embedding(text: str, is_query: bool = False) -> np.ndarray:
    if is_query:
        return dense_model.encode(text, prompt_name="query", normalize_embeddings=True)
    return dense_model.encode(text, normalize_embeddings=True)


def get_dense_embeddings_batch(texts: list, is_query: bool = False, batch_size: int = 64) -> np.ndarray:
    if is_query:
        return dense_model.encode(texts, prompt_name="query", normalize_embeddings=True,
                                  batch_size=batch_size, show_progress_bar=True)
    return dense_model.encode(texts, normalize_embeddings=True, batch_size=batch_size, show_progress_bar=True)

Loading Snowflake/snowflake-arctic-embed-l-v2.0 on MPS...


Loading weights: 100%|██████████| 391/391 [00:00<00:00, 1633.57it/s, Materializing param=pooler.dense.weight]                               


Embedding dimension: 1024
Model device: mps:0


In [40]:
print(f"Generating snowflake-arctic-embed-l-v2.0 embeddings for {len(cleaned_documents):,} documents...")
print("(~10-15 min on CPU / ~3-5 min on Apple Silicon MPS)")

dense_embeddings = get_dense_embeddings_batch(cleaned_documents, is_query=False, batch_size=64)
print(f"\nEmbeddings shape: {dense_embeddings.shape}")

np.save("embeddings_snowflake.npy", dense_embeddings)
print("Saved to embeddings_snowflake.npy")

Generating snowflake-arctic-embed-l-v2.0 embeddings for 65,224 documents...
(~10-15 min on CPU / ~3-5 min on Apple Silicon MPS)


Batches: 100%|██████████| 1020/1020 [39:27<00:00,  2.32s/it]



Embeddings shape: (65224, 1024)
Saved to embeddings_snowflake.npy


### 9.4 Build FAISS Dense Index + BM25 Sparse Index
Two complementary retrieval paths:
- **Dense (FAISS)** — semantic similarity via BGE embeddings
- **Sparse (BM25)** — keyword/term frequency matching

In [41]:
# --- Dense FAISS index ---
dense_dim = dense_embeddings.shape[1]
enhanced_index = faiss.IndexFlatIP(dense_dim)
enhanced_index.add(dense_embeddings.astype("float32"))
print(f"FAISS index: {enhanced_index.ntotal:,} vectors, dim={dense_dim}")

# --- BM25 sparse index ---
print("Building BM25 index...")
tokenized_docs = [doc.lower().split() for doc in cleaned_documents]
bm25_index = BM25Okapi(tokenized_docs)
print(f"BM25 index: {len(tokenized_docs):,} documents")

FAISS index: 65,224 vectors, dim=1024
Building BM25 index...
BM25 index: 65,224 documents


### 9.5 Cross-Encoder Reranker — zerank-1-small
`zeroentropy/zerank-1-small` is the exact reranker used by all top-4 leaderboard entries. It's open-source (Apache 2.0) and outperforms Cohere Rerank-3.5 on BEIR benchmarks.

A cross-encoder scores (query, document) pairs jointly — far more accurate than bi-encoder cosine similarity, but only feasible on a small candidate pool. We apply it to the top-50 fused candidates from hybrid retrieval.

In [42]:
%pip install -q protobuf

import torch
from sentence_transformers import CrossEncoder

# is_accelerate_available() uses @lru_cache — clear it so the post-install
# state of accelerate is detected fresh on the next call.
from transformers.utils import is_accelerate_available
is_accelerate_available.cache_clear()
print(f"accelerate available: {is_accelerate_available()}")

# Remove any leftover global default device from previous cell runs.
torch.set_default_device(None)

RERANKER_MODEL = "zeroentropy/zerank-1-small"
print(f"Loading reranker: {RERANKER_MODEL} on CPU...")
# local_files_only=True avoids ConnectErrors — model is already cached locally.
reranker = CrossEncoder(
    RERANKER_MODEL,
    trust_remote_code=True,
    device="cpu",
    local_files_only=True,
)
print("Reranker loaded.")

Note: you may need to restart the kernel to use updated packages.
accelerate available: True
Loading reranker: zeroentropy/zerank-1-small on CPU...


Loading weights: 100%|██████████| 310/310 [00:00<00:00, 1714.72it/s, Materializing param=model.norm.weight]                              
Qwen3ForSequenceClassification LOAD REPORT from: zeroentropy/zerank-1-small
Key          | Status  | 
-------------+---------+-
score.weight | MISSING | 

Notes:
- MISSING	:those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Reranker loaded.


### 9.6 Hybrid Search + Rerank Pipeline
Three-stage retrieval:
1. **Retrieve** top-100 from Dense (FAISS) and top-100 from BM25
2. **Fuse** via Reciprocal Rank Fusion (RRF)
3. **Rerank** top-50 fused candidates with cross-encoder → return top-K

In [43]:
# Prerequisites check — run this before the hybrid search function
_missing = []
if "dense_model"       not in dir(): _missing.append("dense_model       → run cell 9.3 (Load snowflake model)")
if "get_dense_embedding" not in dir(): _missing.append("get_dense_embedding → run cell 9.3 (Load snowflake model)")
if "cleaned_documents" not in dir(): _missing.append("cleaned_documents → run cell 9.2 (Text cleaning)")
if "enhanced_index"    not in dir(): _missing.append("enhanced_index    → run cell 9.4 (Build FAISS + BM25 indices)")
if "bm25_index"        not in dir(): _missing.append("bm25_index        → run cell 9.4 (Build FAISS + BM25 indices)")
if "reranker"          not in dir(): _missing.append("reranker          → run cell 9.5 (Load zerank-1-small)")
if "corpus_data"       not in dir(): _missing.append("corpus_data       → run cell 9.7 baseline eval (defines corpus_data)")

if _missing:
    raise RuntimeError(
        "Missing prerequisites — please run these cells first:\n  " + "\n  ".join(_missing)
    )
print("All prerequisites satisfied. Ready to build hybrid search pipeline.")

All prerequisites satisfied. Ready to build hybrid search pipeline.


In [ ]:
def hybrid_search_rerank(
    query: str,
    top_k: int = 10,
    dense_candidates: int = 100,
    bm25_candidates: int = 100,
    rerank_pool: int = 50,
    rrf_k: int = 60,
):
    """
    Stage 1: Dense + BM25 retrieval
    Stage 2: Reciprocal Rank Fusion
    Stage 3: Cross-encoder reranking
    """
    # --- Stage 1A: Dense retrieval ---
    q_emb = get_dense_embedding(query, is_query=True).astype("float32").reshape(1, -1)
    dense_scores, dense_indices = enhanced_index.search(q_emb, dense_candidates)

    # --- Stage 1B: BM25 retrieval ---
    bm25_scores = bm25_index.get_scores(query.lower().split())
    bm25_top = np.argsort(bm25_scores)[::-1][:bm25_candidates]

    # --- Stage 2: Reciprocal Rank Fusion ---
    rrf_scores = defaultdict(float)
    for rank, idx in enumerate(dense_indices[0]):
        rrf_scores[int(idx)] += 1.0 / (rrf_k + rank + 1)
    for rank, idx in enumerate(bm25_top):
        rrf_scores[int(idx)] += 1.0 / (rrf_k + rank + 1)

    fused = sorted(rrf_scores.items(), key=lambda x: x[1], reverse=True)[:rerank_pool]
    candidate_indices = [idx for idx, _ in fused]

    # --- Stage 3: Cross-encoder reranking ---
    pairs = [(query, cleaned_documents[idx]) for idx in candidate_indices]
    rerank_scores = reranker.predict(pairs)

    scored = sorted(zip(candidate_indices, rerank_scores), key=lambda x: x[1], reverse=True)

    results = []
    for idx, score in scored[:top_k]:
        doc = corpus_data[idx]
        results.append({
            "id": doc["id"],
            "text": doc["text"],
            "title": doc["title"],
        })
    return results


# Quick sanity test
test_result = hybrid_search_rerank("How do I set up AirSync?", top_k=3)
for i, r in enumerate(test_result, 1):
    print(f"[{i}] {r['title']}")
    print(f"    {r['text'][:120]}...")
    print()

Loading weights: 100%|██████████| 310/310 [00:00<00:00, 2233.24it/s, Materializing param=model.norm.weight]                              


### 9.7 Evaluate Enhanced Pipeline vs Baseline
Run the full enhanced pipeline on all 291 annotated queries and compare head-to-head.

In [2]:
print("Evaluating enhanced pipeline on annotated queries...")
print("(~2-4 min: 291 queries × reranking 50 candidates each)\n")

enhanced_eval_results = []
for item in tqdm(annotated_queries["train"], desc="Enhanced eval"):
    retrievals = hybrid_search_rerank(item["query"], top_k=10)
    enhanced_eval_results.append({
        "query_id": item["query_id"],
        "query": item["query"],
        "retrievals": retrievals,
    })

enhanced_metrics = evaluate_retrieval(enhanced_eval_results, ground_truth)

# Leaderboard reference targets (open-source entries, in %)
LEADERBOARD = {
    "Rank 3 OS (snowflake+zerank)": {"Precision@10": 26.63, "Recall@10": 36.09},
    "Rank 4 OS (GTE-Qwen2+zerank)": {"Precision@10": 26.85, "Recall@10": 35.08},
}

print("\n" + "=" * 80)
print(f"{'METRIC':<16} {'BASELINE':>10} {'ENHANCED':>10} {'DELTA':>8}  {'Rank3 OS':>10} {'Gap to R3':>10}")
print("=" * 80)
for k in sorted(baseline_metrics.keys()):
    b = baseline_metrics[k]
    e = enhanced_metrics[k]
    delta = e - b
    r3 = LEADERBOARD["Rank 3 OS (snowflake+zerank)"].get(k, None)
    gap_str = f"{e - r3:+.2f}" if r3 is not None else "    —"
    r3_str  = f"{r3:.2f}%" if r3 is not None else "   —"
    arrow = "+" if delta > 0 else ""
    print(f"  {k:<14} {b:>9.2f}% {e:>9.2f}% {arrow}{delta:>7.2f}  {r3_str:>10} {gap_str:>10}")
print("=" * 80)
print("\nAll values in % — same scale as the leaderboard.")
print("Leaderboard evaluates on TEST QUERIES (92 held-out); this eval uses annotated_queries (291) as proxy.")

Evaluating enhanced pipeline on annotated queries...
(~2-4 min: 291 queries × reranking 50 candidates each)



NameError: name 'tqdm' is not defined

### 9.8 Generate Final Submission — Test Queries
Run the enhanced pipeline on all 92 held-out test queries and save in the required format.

In [ ]:
print("Running enhanced pipeline on 92 test queries...")

enhanced_test_results = []
for item in tqdm(test_queries["test"], desc="Test queries"):
    retrievals = hybrid_search_rerank(item["query"], top_k=10)
    enhanced_test_results.append({
        "query_id": item["query_id"],
        "query": item["query"],
        "retrievals": retrievals,
    })

OUTPUT_FILE = "test_queries_results_enhanced.json"
with open(OUTPUT_FILE, "w") as f:
    json.dump(enhanced_test_results, f, indent=2)

print(f"\n{'=' * 50}")
print(f"Results saved to {OUTPUT_FILE}")
print(f"  Queries:    {len(enhanced_test_results)}")
print(f"  Per query:  10 retrievals")
print(f"{'=' * 50}")

results_df = pd.DataFrame(enhanced_test_results)
results_df.to_parquet("test_queries_results_enhanced.parquet", index=False)
print("Also saved to test_queries_results_enhanced.parquet")

Running enhanced pipeline on 92 test queries...


Test queries: 100%|██████████| 92/92 [00:51<00:00,  1.80it/s]


Results saved to test_queries_results_enhanced.json
  Queries:    92
  Per query:  10 retrievals
Also saved to test_queries_results_enhanced.parquet


In [ ]:
# Preview first enhanced result
print("Sample enhanced result:\n")
print(json.dumps(enhanced_test_results[0], indent=2, default=str)[:1500])

Sample enhanced result:

{
  "query_id": "a97f93d2-410a-431f-ae9a-1e23ed35d74c",
  "query": "end customer organization name not appearing in ticket or conversation",
  "retrievals": [
    {
      "id": "ART-3207_KNOWLEDGE_NODE-53",
      "text": "can click **Accept invite**.\\n\\n4. Verify configuration\\n-----------------------\\n\\n1. Send an email to the configured support email address.\\n2. Confirm that a ticket or conversation is created in the DevRev app.\\n3. Add a comment to the ticket or conversation and verify that it is received at the email address that sent the original message.\\n\\nTroubleshooting deliverability\\n\\nWhile the DevRev app ensures that emails are sent, the deliverability of the emails to the end customers is",
      "title": "Email snap-in configuration | Email | Integrate | Snap-ins | DevRev"
    },
    {
      "id": "ART-6174_KNOWLEDGE_NODE-29",
      "text": "rules\\n\\nWorkflows enable seamless handovers from automated conversations to your support te

### 9.9 Load Saved Enhanced Embeddings (Optional)
Skip re-embedding if you've already generated and saved `embeddings_bge_large.npy`.

In [ ]:
# Uncomment to load pre-saved embeddings instead of re-generating:
# dense_embeddings = np.load("embeddings_snowflake.npy")
# print(f"Loaded embeddings: {dense_embeddings.shape}")
# Then re-run cells 9.4 onward (skip 9.3 embedding generation)